enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,267 @@
+    # Prevent upgrades to OpenShift Version 4.9
+    operators.coreos.com/maxOpenShiftVersion: 4.8.0 
+```
+
+> NOTE: OLM should allow users to specify Major.Minor.Patch versions in this field BUT OLM can only prevent Major.Minor OpenShift upgrades. By allowing Major.Minor.Patch versions to be specified, OLM will not need to modify the existing annotations if OpenShift ever allows ClusterOperators to prevent patch upgrades.
+
+Assume that the CSV shown above is the only CSV on the Cluster. In this case, OLM would compare the value of the `operators.coreos.com/maxOpenShiftVersion` annotation against the version of the Cluster to decide it's upgradeable status.
+
+Examples can be found below:
+
+| OpenShift Cluster Version   | CSV's Max OpenShift Version | Is Upgradeable?|
+| --------------------------- | --------------------------- | -------------- |
+| 4.6.z                       | 4.8.0                       | True           |
+| 4.7.z                       | 4.8.0                       | True           |
+| 4.8.z                       | 4.8.0                       | False          |
+
+Examples of possible OLM Upgradeable conditions based on installed operators can be found below:
+```yaml=
+
+```
+
+As this is an opt-in feature, there are instances where OLM will be unable to determine if the upgrade is safe:
+
+- The CSV is missing the `operators.coreos.com/maxOpenShiftVersion` annotation
+- The value associated with the `operators.coreos.com/maxOpenShiftVersion` is not a valid Semantic Version (Valid Example: x.y.z).
+
+OLM will report the following upgradeable conditions based on the collection of CSV present on cluster:
+
+| Key                | Meaning                                                                                            |
+| -------------------| -------------------------------------------------------------------------------------------------- |
+| Upgradeable    CSV | A CSV with a valid upgradeable annotation that will run on the next Minor version of OpenShift     |
+| Unupgradeable  CSV | A CSV with a valid upgradeable annotation that will not run on the next Minor version of OpenShift |
+| Undeterminable CSV | A CSV that does not include a valid upgradeable annotation                                         |
+
+| Upgradeable CSV | Indeterminate CSV | Unupgradeable CSV | Is Upgradeable | Upgradeable Message |
+| --------------- | ----------------- | ----------------- | -------------- | ------------------------------------------------------------------------- |
+| X               |                   |                   | True           | All operators will run on the next Minor OpenShift Version                |
+| X               | X                 |                   | True           | One or more operator(s) might not run on the next Minor OpenShift Version |
+| X               | X                 | X                 | False          | Not all operators will run on the next Minor OpenShift Version            |
```

Comment:
not sure why multiple columns have Xs here.  If there's a single "indeterminate" CSV, it doesn't matter if there are any "upgradeable" CSVs.

similarly if there's a signle unupgradeable CSV, it doesn't matter if there are any indeterminate or upgradeable CSVs.

Instead of Xs, should those boxes have a "don't care" value?

Reply:
> not sure why multiple columns have Xs here. 

X was meant to indicate the presence of a CSV of that type.

>  If there's a single "indeterminate" CSV, it doesn't matter if there are any "upgradeable" CSVs.

OLM would not prevent an upgrade if one or more "indeterminate" CSVs are present provided no "unupgradeable" csv exists.

As long as you agree that that's the logic that we want I can work on improving the table to convey that point.

Reply:
> X was meant to indicate the presence of a CSV of that type.

yeah i don't think that's the best way to represent the logic, because the presence of an upgradeable CSV literally does not matter (and the presence of an indeterminate CSV does not matter if you have a non-upgradeable CSV present).  Hence my suggestion of "don't care" as the value to use.  As written the table doesn't explicitly state what happens if "there are no upgradeable CSVs and some indeterminate CSVs" because it only states what happens if "there are upgradeable CSVs and indeterminate CSVs"  (second row).


> OLM would not prevent an upgrade if one or more "indeterminate" CSVs are present provided no "unupgradeable" CSVs exists.

right.  but it doesn't matter if there are (or are not) "upgradeable" CSVs for that case.

Reply:
> yeah i don't think that's the best way to represent the logic, because the presence of an upgradeable CSV literally does not matter (and the presence of an indeterminate CSV does not matter if you have a non-upgradeable CSV present).

I disagree, the presence of an indeterminate CSV impacts the ClusterOperator message.

>  Hence my suggestion of "don't care" as the value to use. As written the table doesn't explicitly state what happens if "there are no upgradeable CSVs and some indeterminate CSVs" because it only states what happens if "there are upgradeable CSVs and indeterminate CSVs" (second row).


Perhaps instead of a table we should make it clear that:

- "Unupgradeable" operators set Degraded to true
- "Indeterminable" CSVs does not change Degraded to true but will update the message.
- -"Upgradeable" CSVs that reports a maxOpenShiftVersion

Reply:
Updated the table to reflect changes that @bparees and I discussed offline:

```

| Upgradeable CSV Present | Indeterminate CSV Present | Non-Upgradeable CSV Present | Status          | Message            |
| ----------------------- | ------------------------- | --------------------------- |---------------- | ------------------ |
| Yes                     | No                        | No                          | Upgradeable     | Ready for upgrade  |
| Don't Care              | Yes                       | No                          | Upgradeable     | Warning            |
| Don't Care              | Don't Care                | Yes                         | Not Upgradeable | Upgrade is blocked |
```